src/MySqlConnector/MySqlClient/CommandExecutors/StoredProcedureCommandExecutor.cs

```diff
@@ -0,0 +1,92 @@
+		{
+			m_command = command;
+		}
+
+		public async Task<int> ExecuteNonQueryAsync(string commandText, MySqlParameterCollection parameterCollection, IOBehavior ioBehavior,
+			CancellationToken cancellationToken)
+		{
+			await ExecuteReaderAsync(commandText, parameterCollection, CommandBehavior.Default, ioBehavior, cancellationToken).ConfigureAwait(false);
+			return m_affectedRows;
+		}
+
+		public async Task<object> ExecuteScalarAsync(string commandText, MySqlParameterCollection parameterCollection, IOBehavior ioBehavior,
+			CancellationToken cancellationToken)
+		{
+			await ExecuteReaderAsync(commandText, parameterCollection, CommandBehavior.Default, ioBehavior, cancellationToken).ConfigureAwait(false);
+			return m_returnValue;
+		}
+
+		public async Task<DbDataReader> ExecuteReaderAsync(string commandText, MySqlParameterCollection parameterCollection, CommandBehavior behavior,
+			IOBehavior ioBehavior, CancellationToken cancellationToken)
+		{
+			var inParams = new MySqlParameterCollection();
+			var outParams = new MySqlParameterCollection();
+			var argParamNames = new List<string>();
+			var outParamNames = new List<string>();
+			MySqlParameter returnParameter = null;
+			for (var i = 0; i < parameterCollection.Count; i++)
+			{
+				var parameter = parameterCollection[i];
+				if (parameter.Direction == ParameterDirection.Input || parameter.Direction == ParameterDirection.InputOutput)
+				{
+					inParams.Add(parameter);
+				}
+				if (parameter.Direction == ParameterDirection.Output || parameter.Direction == ParameterDirection.InputOutput)
+				{
+					outParams.Add(parameter);
+					outParamNames.Add(parameter.ParameterName);
+				}
+
+				if (parameter.Direction == ParameterDirection.ReturnValue)
+					returnParameter = parameter;
+				else
+					argParamNames.Add(parameter.ParameterName);
+			}
+
+			commandText = "CALL " + commandText + "(" + string.Join(", ", argParamNames) +")";
+			using (var reader = await DoExecuteReaderAsync(commandText, inParams, CommandBehavior.SingleResult | CommandBehavior.SingleRow, ioBehavior, cancellationToken).ConfigureAwait(false))
+			{
+				do
+				{
+					if (await reader.ReadAsync(ioBehavior, cancellationToken).ConfigureAwait(false))
+						m_returnValue = reader.GetValue(0);
+				} while (await reader.NextResultAsync(ioBehavior, cancellationToken).ConfigureAwait(false));
+				m_affectedRows = reader.RecordsAffected;
+			}
+
+			if (returnParameter != null)
+				returnParameter.Value = m_returnValue;
+
+			if (outParams.Count > 0)
+			{
+				commandText = "SELECT " + string.Join(", ", outParamNames);
+				using (var reader = await DoExecuteReaderAsync(commandText, new MySqlParameterCollection(), CommandBehavior.Default, ioBehavior, cancellationToken).ConfigureAwait(false))
+				{
+					await reader.ReadAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
+					for (var i = 0; i < outParams.Count; i++)
+						outParams[i].Value = reader.GetValue(i);
+				}
+			}
+
+			return null;
```

Comment:
I think `MySql.Data` may actually return a reader here.  How could we return a reader since it is used up when getting the return value?